torch/onnx/_backend/core.py

```diff
@@ -0,0 +1,852 @@
+                output_names=output_names,
+                output_value_infos=tuple(output for output in onnx_model.graph.output),
+                input_devices=input_devices,
+                output_devices=output_devices,
+                example_outputs=prim_outputs,
+            )
+
+            self._all_ort_execution_info.cache_session_execution_info(
+                graph_module, execution_info_per_session
+            )
+
+        self.execution_count += 1
+
+        if isinstance(prim_outputs, tuple):
+            assert all(isinstance(elem, torch.Tensor) for elem in prim_outputs)
+            # ORT always returns a tuple of outputs. If the original is a tuple, just returning
+            # ORT output is ok.
+            _nvtx_range_push("run_onnx_session_with_ortvaluevector")
+            onnx_outputs = _run_onnx_session_with_ortvaluevector(
+                onnx_session,
+                input_names,
+                args,
+                input_devices,
+                output_names,
+                prim_outputs,
+                output_devices,
+                self.preallocate_output,
+            )
+            _nvtx_range_pop()
+            if self._assert_allclose_to_baseline:
+                # Compute baseline.
+                baseline_outputs = torch._prims.executor.execute(
+                    graph_module, *args, executor="aten"
+                )
+                # Ensure every output tensor is close to the corresponding baseline.
+                for onnx_output, baseline_output in zip(onnx_outputs, baseline_outputs):
+                    _assert_allclose_with_detailed_error_message(
+                        onnx_output, baseline_output
+                    )
+            return onnx_outputs
+        else:
+            assert isinstance(prim_outputs, torch.Tensor)
+            # ORT always returns a tuple of outputs. If the original output is a tensor,
+            # ORT output's first element must be extracted and returned. Otherwise, type
+            # mismatch may happen in downstream computation.
+            onnx_outputs = _run_onnx_session_with_ortvaluevector(
+                onnx_session,
+                input_names,
+                args,
+                input_devices,
+                output_names,
+                (prim_outputs,),
+                output_devices,
+                self.preallocate_output,
+            )
+            assert len(onnx_outputs) == 1
+            if self._assert_allclose_to_baseline:
+                # Compute baseline.
+                baseline_outputs = torch._prims.executor.execute(
+                    graph_module, *args, executor="aten"
+                )
+                # Ensure output tensor is close to the corresponding baseline.
+                _assert_allclose_with_detailed_error_message(
+                    onnx_outputs[0], baseline_outputs
+                )
+            return onnx_outputs[0]
```

Comment:
I have no plan to avoid repeated codes since it will dramatically reduce readability for me. You can see for corresponding function calls in this if-else block, they are different at least at one argument. If you want to merge them, you will need a flag like
```
        if isinstance(prim_outputs, tuple) and all(isinstance(elem, torch.Tensor) for elem in prim_outputs):
            original_graph_module_output_is_a_tuple_of_tensors = True
        elif isinstance(prim_outputs, torch.Tensor):
            original_graph_module_output_is_a_tuple_of_tensors = False
        else:
            raise RuntimeError(
                f"prim_outputs must be either a tuple of tensors or a tensor. prim_outputs={prim_outputs}"
            )
```

and use it to create an if-else for each function call (_run_onnx_session_with_ortvaluevector, torch.testing.assert_close). It will end up with independent 3 if-else blocks, which read very bad. You probably can hide some logics inside `_run_onnx_session_with_ortvaluevector` but I only hide when there is no way to avoid.

Reply:
Below is easier to maintain and I don't think it is "dramatically reducing readability".
```python
# ORT always returns a tuple of outputs. If the original output is a tensor,
# ORT output's first element must be extracted and returned. Otherwise, type
# mismatch may happen in downstream computation.
is_single_tensor_output = isinstance(prim_outputs, torch.Tensor)
normalized_prim_outputs = (prim_outputs,) if is_single_tensor_output else prim_outputs
assert isinstance(normalized_prim_outputs, tuple)
assert all(isinstance(elem, torch.Tensor) for elem in normalized_prim_outputs)

_nvtx_range_push("run_onnx_session_with_ortvaluevector")
onnx_outputs = _run_onnx_session_with_ortvaluevector(
    onnx_session,
    input_names,
    args,
    input_devices,
    output_names,
    normalized_prim_outputs,
    output_devices,
    self.preallocate_output,
)
_nvtx_range_pop()
if self._assert_allclose_to_baseline:
    # Compute baseline.
    baseline_outputs = torch._prims.executor.execute(
        graph_module, *args, executor="aten"
    )
    normalized_baseline_ouptuts = (baseline_outputs,) if is_single_tensor_output else baseline_outputs
    # Ensure every output tensor is close to the corresponding baseline.
    for onnx_output, baseline_output in zip(onnx_outputs, normalized_baseline_ouptuts):
        torch.testing.assert_close(onnx_output, baseline_output)
return onnx_outputs[0] if is_single_tensor_output else onnx_outputs
```

Reply:
Done.